### PR TITLE
Moved run_all_in_graph_and_eager_mode in basic decoder

### PR DESCRIPTION
**Pull request by autoih**
cc @gabrieldemarmiesse. Dividing into several parts since it's large, and not sure if removing `use_gpu` is a good idea. The reason is using gpu seems not helping a lot in LSTM, only if it's self-attention(transformer) based. May need your suggestions.

### CODE REVIEW

**Comment by gabrieldemarmiesse** [comment]
About the use of gpu for testing going forward, unless we're testing some code with a custom op, we need a strong reason to run in both cpu and gpu mode.

File: tensorflow_addons/seq2seq/tests/basic_decoder_test.py [hunk 1]
```diff
@@ -24,372 +25,301 @@
 from tensorflow_addons.seq2seq import sampler as sampler_py
 
 
-class BasicDecoderTest(tf.test.TestCase, parameterized.TestCase):
-    """Unit test for basic_decoder.BasicDecoder."""
-
-    @parameterized.named_parameters(
-        ("use_output_layer", True), ("without_output_layer", False)
+def test_step_with_training_helper_output_layer(use_output_layer=True):
```
Comment (gabrieldemarmiesse) [inline tensorflow_addons/seq2seq/tests/basic_decoder_test.py:28]:
```suggestion
@pytest.mark.parametrize("use_output_layer", [True, False])
def test_step_with_training_helper_output_layer(use_output_layer):
```

File: tensorflow_addons/seq2seq/tests/basic_decoder_test.py [hunk 2]
```diff
@@ -24,372 +25,301 @@
...
-            output_size = my_decoder.output_size
-            output_dtype = my_decoder.output_dtype
-            self.assertEqual(
-                basic_decoder.BasicDecoderOutput(cell_depth, tf.TensorShape([])),
-                output_size,
-            )
-            self.assertEqual(
-                basic_decoder.BasicDecoderOutput(tf.float32, tf.int32), output_dtype
-            )
+@pytest.mark.parametrize("use_mask", [True, False])
```
Comment (gabrieldemarmiesse) [inline tensorflow_addons/seq2seq/tests/basic_decoder_test.py:108]:
```suggestion
@pytest.mark.parametrize("use_mask", [True, False, None])
```

File: tensorflow_addons/seq2seq/tests/basic_decoder_test.py (hunk 1 above)
Comment (autoih) [reply tensorflow_addons/seq2seq/tests/basic_decoder_test.py:28]: Oh, yeah, it also tests none use_output_layer. Thanks @gabrieldemarmiesse
